src/langchain_google_firestore/document_loader.py

```diff
@@ -0,0 +1,199 @@
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+import itertools
+from typing import TYPE_CHECKING, Any, Iterator, List, Optional
+
+import more_itertools
+from google.cloud import firestore
+from google.cloud.firestore import DocumentReference
+from google.cloud.firestore_v1.services.firestore.transports.base import (
+    DEFAULT_CLIENT_INFO,
+)
+from langchain_community.document_loaders.base import BaseLoader
+from langchain_core.documents import Document
+
+from .utility.document_converter import DocumentConverter
+
+DEFAULT_FIRESTORE_DATABASE = "(default)"
+USER_AGENT = "langchain-google-firestore-python"
+WRITE_BATCH_SIZE = 500
+
+if TYPE_CHECKING:
+    from google.cloud.firestore import Client, CollectionGroup, DocumentReference, Query
+
+
+class FirestoreLoader(BaseLoader):
+    def __init__(
+        self,
+        source: Query | CollectionGroup | DocumentReference | str,
+        page_content_fields: List[str] = None,
+        metadata_fields: Optional[List[str]] = None,
+        client: Optional[Client] = None,
+    ) -> None:
+        """Document Loader for Google Cloud Firestore.
+        Args:
+            source: The source to load the documents. It can be an instance of Query,
+                CollectionGroup, DocumentReference or the single `/`-delimited path to
+                a Firestore collection.
+            page_content_fields: The document field names to write into the `page_content`.
+                If an empty or None list is provided all fields will be written into
+                `page_content`. When only one field is provided only the value is written.
+            metadata_fields: The document field names to write into the `metadata`.
+                By default it will write all fields that are not in `page_content` into `metadata`.
+            client: Client for interacting with the Google Cloud Firestore API.
+        """
+        if client:
+            self.client = client
+            self.client._user_agent = USER_AGENT
+        else:
+            client_info = DEFAULT_CLIENT_INFO
+            client_info.user_agent = USER_AGENT
+            self.client = firestore.Client(client_info=client_info)
+        self.source = source
+        self.page_content_fields = page_content_fields
+        self.metadata_fields = metadata_fields
+
+    def load(self) -> List[Document]:
+        """Load Documents."""
+        return list(self.lazy_load())
+
+    def lazy_load(self) -> Iterator[Document]:
+        """A lazy loader for Documents."""
+        if isinstance(self.source, DocumentReference):
```

Comment:
How do we handle CollectionGroup? If different type needs special handle, this can be an ENUM.

Reply:
The ColllectionGroup has the same methods to do the query. So we just keep it as query since we can do query.stream() 
There's a unit tests for this. 

About ENUM, it's str, DocumentReference or other. But I'm not quite sure how to set Enum with class types; did a quick search online but didn't find checking the class type. Unless I compare strings.

Reply:
I was referring something like: 

class Source(Enum):
  DOCUMENT_REFERENCE = 1
  DOCUMENT_NAME = 2
  QUERY = 3 
  COLLECTION_GROUP = 4

to avoid many `isinstance()` condtiions. https://docs.python.org/3/library/enum.html.

Reply:
I'm sorry I'm can't find references of how to check the instance with an Enum.  Since I have the object reference I still need to figure out the class type.

So for example I can use    `isinstance(a, Source)` from your example to tell is an enum of Source.  
But if I have a DocumentReference -- I need to do some mapping to this Enum , right?

I was checking locally and I can't get a string from   __class__  or __str__ that would be simpler to then use with the enum approach